Add CSS `@custom-media` at-rule (#42075)
mdn/content: files/en-us/web/css/reference/at-rules/@custom-media/index.md

---

> title: "@custom-media"

Reviewer: A link to this page also needs to be added to the CSS media queries module reference page, in the relevant section:

https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Media_queries#at-rules_and_descriptors

slug: Web/CSS/Reference/At-rules/@custom-media
page-type: css-at-rule
status:
  - experimental
browser-compat: css.at-rules.custom-media
---

{{CSSRef}}{{SeeCompatTable}}

The **`@custom-media`** CSS [at-rule](/en-US/docs/Web/CSS/At-rule) defines aliases for longer or complex [media queries](/en-US/docs/Web/CSS/CSS_media_queries). Instead of using the same hardcoded [media features](/en-US/docs/Web/CSS/@media#media_features) {{CSSxRef("@media")}} query list multiple times in multiple CSS files, a `@custom-media` query is defined once, in a single location, and can then be used everywhere.

## Syntax

```css
@custom-media --media-query-name (max-width: 1200px);
@custom-media --media-query-name (max-width: 1200px), (orientation: portrait);
```

### Values

- `<extension-name>`
  - : A [`<dashed-ident>`](/en-US/docs/Web/CSS/dashed-ident); the name identifying the custom media.
- `<media-query-list>`

>   - : A comma-separated [list of `<media-query>` values](/en-US/docs/Web/CSS/@media#description).

Reviewer: If we're going to use `<media-query>` here, then I think it would be good to mention `<media-query>` in the Description section of the `@media` page. Maybe something like:

"A media query's `<media-query-list>` is a comma-separated list of one or more `<media-query>` values. Each `<media-query>` is comprised of <media-type>s, <media-feature>s, and logical operators."

## Description

When designing documents that use media queries, an {{CSSxRef("@media")}} rule may be used in multiple places. Repeating the same media query values multiple times can lead to errors and maintenance issues. If a change needs to be made, every occurrence must be found and updated in the same way, which may be hard to detect, especially when media queries are used in multiple files and multiple teams.

> To help address this, the custom media queries at-rule defines named aliases that can be used for long and complex media queries. In this way, a media query used in multiple places can instead be defined as a custom media query, which can be used everywhere. Then, updating the media query requires touching only one line of code.

Reviewer: ```suggestion
To help address this, the `@custom-media` at-rule allows defining named aliases for long, complex media queries. A media query used in multiple places can be defined via a custom media query, and its alias can then be substituted for the media query in all the places it is used. It can then be updated in one place if an update is required.
```

> A `@custom-media` rule can refer to other custom media queries and include the alias name within the feature list. However, ensure the custom media query is not defined in terms of itself or of another custom media query that directly or indirectly refers to it as loops are not valid. Any attempt to define a custom media query with a circular dependency will cause all the custom media queries in the loop to fail.

Author: https://github.com/mdn/content/pull/42075#discussion_r2564581709

If multiple `@custom-media` rules are declared using the same dashed-identifier, the last declaration defined is used, with all previous declarations using the same name being ignored.

## Formal syntax

{{csssyntax}}

## Examples

> ### Updating multiple responsive breakpoints

Reviewer: ```suggestion
### Updating multiple media queries
```

The `@custom-media` at-rule is used on responsive websites which use a particular breakpoint in several places:

```css
@custom-media --narrow-window (max-width: 32em);

@media (--narrow-window) {
}

@media (--narrow-window) and (hover) {
}

@media (--narrow-window) and (orientation: portrait) {
}
```

If the breakpoint changes, by changing one line of code, a single media feature value, all media queries using this custom media query alias in all CSS files are updated accordingly.

### Grouping multiple responsive breakpoints

The `@custom-media` at-rule is used to set multiple breakpoints in a single place:

```css
@custom-media --mobile-screen (max-width: 480px);
@custom-media --tablet-screen (max-width: 768px);
@custom-media --laptop-screen (max-width: 1024px);
@custom-media --desktop-screen (max-width: 1440px);
@custom-media --widescreen (min-width: 1441px);
```

> By grouping all the breakpoints in a single location, it's easier to control the responsive design.

Reviewer: ```suggestion
Grouping all the breakpoints in a single location makes it easier to maintain the responsive design.
```

> ### Complex breakpoints with a list of queries

Author: Done, added en example with multiple files.

To help address this, the custom media queries at-rule defines named aliases that can be used for long and complex media queries. In this way, a media query used in multiple places can instead be defined as a custom media query, which can be used everywhere. Then, updating the media query requires touching only one line of code.

A `@custom-media` rule can refer to other custom media queries and include the alias name within the feature list. However, ensure the custom media query is not defined in terms of itself or of another custom media query that directly or indirectly refers to it as loops are not valid. Any attempt to define a custom media query with a circular dependency will cause all the custom media queries in the loop to fail.

If multiple `@custom-media` rules are declared with the same `<dashed-ident>` name, the last one in the source order is used, with all previous declarations using the same name being ignored.

## Formal syntax

{{csssyntax}}

## Examples

### Updating multiple media queries

In this example, the `@custom-media` at-rule is used on a responsive website that uses a particular breakpoint in several places:

```css

> @custom-media --narrow-window (max-width: 32em);

Author: @yisibl I agree. Thank you for the feedback!

---
title: "@custom-media"
slug: Web/CSS/Reference/At-rules/@custom-media
page-type: css-at-rule
status:
  - experimental
browser-compat: css.at-rules.custom-media
---

{{CSSRef}}{{SeeCompatTable}}

The **`@custom-media`** CSS [at-rule](/en-US/docs/Web/CSS/Reference/At-rules) defines aliases for long or complex [media queries](/en-US/docs/Web/CSS/Guides/Media_queries). Instead of repeating the same hardcoded `<media-query-list>` in multiple [`@media`](/en-US/docs/Web/CSS/Reference/At-rules/@media) at-rules, it can be defined once in a `@custom-media` at-rule and referenced throughout the stylesheet whenever needed.

## Syntax

```css
@custom-media <extension-name> <media-query-list>;

@custom-media --media-query-name (width < 1200px);
@custom-media --media-query-name (width < 1200px), (orientation: portrait);
```

### Values

- `<extension-name>`
  - : A [`<dashed-ident>`](/en-US/docs/Web/CSS/Reference/Values/dashed-ident); the name identifying the custom media query.
- `<media-query-list>` | `true` | `false`

>   - `<media-query-list>`

Reviewer: Add a line about this along the lines of "The `<media-query-list>` or keyword represented by the custom media query. Possible values are:"

### Evaluating media queries with logical operators

Custom media queries accept the full range of CSS logical operators — `not`, `and`, and `or` (comma-separated or using the `or` keyword).

Because a `@custom-media` value is just a normal `<media-query-list>`, you can combine, invert, or group conditions exactly as you would in a regular `@media` rule.

#### Using the `not` operator

The `not` operator negates an entire media condition. This is useful when you want a rule to apply only when a specific condition is `false`.

```css
@custom-media --no-script not (script);

@media (--no-script) {
}
```

#### Using the `and` operator

The `and` operator lets you combine multiple conditions that must all be `true`.

```css
@custom-media --medium-screen (min-width: 40em) and (max-width: 60em);

@media (--medium-screen) {
}
```

This alias only matches when the viewport is within the specified width range.

#### Using the `or` operator (comma-separated list)

A comma creates a logical `or` operator. The media query matches if any of the listed conditions are `true`.

```css

> @custom-media --screen-or-print screen, print;

Reviewer: Yes, but you still don't show what the actual `or` code equivalent looks like. You show the comma version, but not the `or` version.
